// @elastic/charts/dist/common/iterables.js
"use strict";
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.pipeline = exports.executing = exports.filter = exports.map = exports.filtering = exports.doing = exports.mapping = void 0;
function mapping(fun) {
    return function (iterable) {
        var iterable_1, iterable_1_1, next, e_1_1;
        var e_1, _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _b.trys.push([0, 5, 6, 7]);
                    iterable_1 = __values(iterable), iterable_1_1 = iterable_1.next();
                    _b.label = 1;
                case 1:
                    if (!!iterable_1_1.done) return [3, 4];
                    next = iterable_1_1.value;
                    return [4, fun(next)];
                case 2:
                    _b.sent();
                    _b.label = 3;
                case 3:
                    iterable_1_1 = iterable_1.next();
                    return [3, 1];
                case 4: return [3, 7];
                case 5:
                    e_1_1 = _b.sent();
                    e_1 = { error: e_1_1 };
                    return [3, 7];
                case 6:
                    try {
                        if (iterable_1_1 && !iterable_1_1.done && (_a = iterable_1.return)) _a.call(iterable_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                    return [7];
                case 7: return [2];
            }
        });
    };
}
exports.mapping = mapping;
function doing(fun) {
    return function (iterable) {
        var iterable_2, iterable_2_1, next;
        var e_2, _a;
        return __generator(this, function (_b) {
            try {
                for (iterable_2 = __values(iterable), iterable_2_1 = iterable_2.next(); !iterable_2_1.done; iterable_2_1 = iterable_2.next()) {
                    next = iterable_2_1.value;
                    fun(next);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (iterable_2_1 && !iterable_2_1.done && (_a = iterable_2.return)) _a.call(iterable_2);
                }
                finally { if (e_2) throw e_2.error; }
            }
            return [2];
        });
    };
}
exports.doing = doing;
function filtering(fun) {
    return function (iterable) {
        var iterable_3, iterable_3_1, next, e_3_1;
        var e_3, _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _b.trys.push([0, 5, 6, 7]);
                    iterable_3 = __values(iterable), iterable_3_1 = iterable_3.next();
                    _b.label = 1;
                case 1:
                    if (!!iterable_3_1.done) return [3, 4];
                    next = iterable_3_1.value;
                    if (!fun(next)) return [3, 3];
                    return [4, next];
                case 2:
                    _b.sent();
                    _b.label = 3;
                case 3:
                    iterable_3_1 = iterable_3.next();
                    return [3, 1];
                case 4: return [3, 7];
                case 5:
                    e_3_1 = _b.sent();
                    e_3 = { error: e_3_1 };
                    return [3, 7];
                case 6:
                    try {
                        if (iterable_3_1 && !iterable_3_1.done && (_a = iterable_3.return)) _a.call(iterable_3);
                    }
                    finally { if (e_3) throw e_3.error; }
                    return [7];
                case 7: return [2];
            }
        });
    };
}
exports.filtering = filtering;
function map(iterable, fun) {
    return mapping(fun)(iterable);
}
exports.map = map;
function filter(iterable, fun) {
    return filtering(fun)(iterable);
}
exports.filter = filter;
function executing(iterable) {
    var iterator = iterable[Symbol.iterator]();
    while (!iterator.next().done) { }
}
exports.executing = executing;
function pipeline(arg) {
    var functions = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        functions[_i - 1] = arguments[_i];
    }
    return functions.reduce(function (iterator, fun) { return fun(iterator); }, arg);
}
exports.pipeline = pipeline;
