// @elastic/eui/lib/components/spacer/spacer.styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.euiSpacerStyles = void 0;

var _react = require("@emotion/react");

var _global_styling = require("../../global_styling");

function _EMOTION_STRINGIFIED_CSS_ERROR__() { return "You have tried to stringify object returned from `css` function. It isn't supposed to be used directly (e.g. as value of the `className` prop), but rather handed to emotion so it can handle it (e.g. as value of `css` prop)."; }

var _ref = process.env.NODE_ENV === "production" ? {
  name: "191xi1n-euiSpacer",
  styles: "flex-shrink:0;label:euiSpacer;"
} : {
  name: "191xi1n-euiSpacer",
  styles: "flex-shrink:0;label:euiSpacer;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};

var euiSpacerStyles = function euiSpacerStyles(_ref2) {
  var euiTheme = _ref2.euiTheme;
  return {
    euiSpacer: _ref,
    // Sizes
    xs: /*#__PURE__*/(0, _react.css)(_global_styling.logicals.height, ":", euiTheme.size.xs, ";;label:xs;"),
    s: /*#__PURE__*/(0, _react.css)(_global_styling.logicals.height, ":", euiTheme.size.s, ";;label:s;"),
    m: /*#__PURE__*/(0, _react.css)(_global_styling.logicals.height, ":", euiTheme.size.base, ";;label:m;"),
    l: /*#__PURE__*/(0, _react.css)(_global_styling.logicals.height, ":", euiTheme.size.l, ";;label:l;"),
    xl: /*#__PURE__*/(0, _react.css)(_global_styling.logicals.height, ":", euiTheme.size.xl, ";;label:xl;"),
    xxl: /*#__PURE__*/(0, _react.css)(_global_styling.logicals.height, ":", euiTheme.size.xxl, ";;label:xxl;")
  };
};

exports.euiSpacerStyles = euiSpacerStyles;
