// @kbn/securitysolution-io-ts-list-types/target_node/src/request/create_rule_exception_item_schema/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createRuleExceptionListItemSchema = void 0;
var t = _interopRequireWildcard(require("io-ts"));
var _securitysolutionIoTsTypes = require("@kbn/securitysolution-io-ts-types");
var _common = require("../../common");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const createRuleExceptionListItemSchema = t.intersection([t.exact(t.type({
  description: _common.description,
  entries: _common.nonEmptyEntriesArray,
  name: _common.name,
  type: _common.exceptionListItemType
})), t.exact(t.partial({
  comments: _common.DefaultCreateCommentsArray,
  // defaults to empty array if not set during decode
  item_id: _securitysolutionIoTsTypes.DefaultUuid,
  // defaults to GUID (uuid v4) if not set during decode
  list_id: t.undefined,
  meta: _common.meta,
  // defaults to undefined if not set during decode
  namespace_type: _common.namespaceType,
  // defaults to 'single' if not set during decode
  os_types: _common.osTypeArrayOrUndefined,
  // defaults to empty array if not set during decode
  tags: _common.tags // defaults to empty array if not set during decode
}))]);
exports.createRuleExceptionListItemSchema = createRuleExceptionListItemSchema;
