// @kbn/securitysolution-io-ts-alerting-types/target_web/src/actions/index.js
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import * as t from 'io-ts';
import { saved_object_attributes } from '../saved_object_attributes';
export var RuleActionGroup = t.string;
export var RuleActionId = t.string;
export var RuleActionTypeId = t.string;

/**
 * Params is an "object", since it is a type of RuleActionParams which is action templates.
 * @see x-pack/plugins/alerting/common/rule.ts
 */

export var RuleActionParams = saved_object_attributes;
export var RuleAction = t.exact(t.type({
  group: RuleActionGroup,
  id: RuleActionId,
  action_type_id: RuleActionTypeId,
  params: RuleActionParams
}));
export var RuleActionArray = t.array(RuleAction);
export var RuleActionCamel = t.exact(t.type({
  group: RuleActionGroup,
  id: RuleActionId,
  actionTypeId: RuleActionTypeId,
  params: RuleActionParams
}));
export var RuleActionArrayCamel = t.array(RuleActionCamel);
