// @elastic/eui/lib/components/timeline/timeline_item.styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.euiTimelineItemStyles = void 0;

var _react = require("@emotion/react");

function _EMOTION_STRINGIFIED_CSS_ERROR__() { return "You have tried to stringify object returned from `css` function. It isn't supposed to be used directly (e.g. as value of the `className` prop), but rather handed to emotion so it can handle it (e.g. as value of `css` prop)."; }

var _ref = process.env.NODE_ENV === "production" ? {
  name: "5bggde-euiTimelineItem",
  styles: "display:flex;position:relative;label:euiTimelineItem;"
} : {
  name: "5bggde-euiTimelineItem",
  styles: "display:flex;position:relative;label:euiTimelineItem;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};

var euiTimelineItemStyles = function euiTimelineItemStyles() {
  return {
    euiTimelineItem: _ref,
    // Vertical alignments
    // These classes are being targeted in timeline.styles.ts
    top: /*#__PURE__*/(0, _react.css)(";label:top;"),
    center: /*#__PURE__*/(0, _react.css)(";label:center;")
  };
};

exports.euiTimelineItemStyles = euiTimelineItemStyles;
