// @kbn/core-saved-objects-server-internal/target_node/src/routes/bulk_delete.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.registerBulkDeleteRoute = void 0;
var _configSchema = require("@kbn/config-schema");
var _utils = require("./utils");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const registerBulkDeleteRoute = (router, {
  coreUsageData
}) => {
  router.post({
    path: '/_bulk_delete',
    validate: {
      body: _configSchema.schema.arrayOf(_configSchema.schema.object({
        type: _configSchema.schema.string(),
        id: _configSchema.schema.string()
      })),
      query: _configSchema.schema.object({
        force: _configSchema.schema.maybe(_configSchema.schema.boolean())
      })
    }
  }, (0, _utils.catchAndReturnBoomErrors)(async (context, req, res) => {
    const {
      force
    } = req.query;
    const usageStatsClient = coreUsageData.getClient();
    usageStatsClient.incrementSavedObjectsBulkDelete({
      request: req
    }).catch(() => {});
    const {
      savedObjects
    } = await context.core;
    const statuses = await savedObjects.client.bulkDelete(req.body, {
      force
    });
    return res.ok({
      body: statuses
    });
  }));
};
exports.registerBulkDeleteRoute = registerBulkDeleteRoute;
