// @kbn/coloring/target_node/src/palettes/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FIXED_PROGRESSION = exports.DEFAULT_RANGE_TYPE = exports.DEFAULT_PALETTE_NAME = exports.DEFAULT_MIN_STOP = exports.DEFAULT_MAX_STOP = exports.DEFAULT_CONTINUITY = exports.DEFAULT_COLOR_STEPS = exports.CUSTOM_PALETTE = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const DEFAULT_PALETTE_NAME = 'positive';
exports.DEFAULT_PALETTE_NAME = DEFAULT_PALETTE_NAME;
const FIXED_PROGRESSION = 'fixed';
exports.FIXED_PROGRESSION = FIXED_PROGRESSION;
const CUSTOM_PALETTE = 'custom';
exports.CUSTOM_PALETTE = CUSTOM_PALETTE;
const DEFAULT_CONTINUITY = 'above';
exports.DEFAULT_CONTINUITY = DEFAULT_CONTINUITY;
const DEFAULT_RANGE_TYPE = 'percent';
exports.DEFAULT_RANGE_TYPE = DEFAULT_RANGE_TYPE;
const DEFAULT_MIN_STOP = 0;
exports.DEFAULT_MIN_STOP = DEFAULT_MIN_STOP;
const DEFAULT_MAX_STOP = 100;
exports.DEFAULT_MAX_STOP = DEFAULT_MAX_STOP;
const DEFAULT_COLOR_STEPS = 5;
exports.DEFAULT_COLOR_STEPS = DEFAULT_COLOR_STEPS;
