// @elastic/eui/lib/components/pagination/pagination.styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.euiPaginationStyles = void 0;

var _react = require("@emotion/react");

var _global_styling = require("../../global_styling");

function _EMOTION_STRINGIFIED_CSS_ERROR__() { return "You have tried to stringify object returned from `css` function. It isn't supposed to be used directly (e.g. as value of the `className` prop), but rather handed to emotion so it can handle it (e.g. as value of `css` prop)."; }

var _ref = process.env.NODE_ENV === "production" ? {
  name: "1dbd2rc-euiPagination__list",
  styles: "display:flex;align-items:baseline;label:euiPagination__list;"
} : {
  name: "1dbd2rc-euiPagination__list",
  styles: "display:flex;align-items:baseline;label:euiPagination__list;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};

var euiPaginationStyles = function euiPaginationStyles(euiThemeContext) {
  var euiTheme = euiThemeContext.euiTheme;
  return {
    // Base
    euiPagination: /*#__PURE__*/(0, _react.css)((0, _global_styling.euiScrollBarStyles)(euiThemeContext), ";display:flex;align-items:center;", (0, _global_styling.logicalCSSWithFallback)('overflow-y', 'hidden'), ";", (0, _global_styling.logicalCSSWithFallback)('overflow-x', 'auto'), ";;label:euiPagination;"),
    // Elements
    euiPagination__compressedText: /*#__PURE__*/(0, _react.css)("display:inline-flex;align-items:center;line-height:1!important;>span{", (0, _global_styling.logicalCSS)('margin-horizontal', euiTheme.size.s), ";font-weight:", euiTheme.font.weight.semiBold, ";&:first-of-type{color:", euiTheme.colors.primaryText, ";}};label:euiPagination__compressedText;"),
    euiPagination__list: _ref
  };
};

exports.euiPaginationStyles = euiPaginationStyles;
