// @elastic/eui/lib/components/overlay_mask/overlay_mask.styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.euiOverlayMaskStyles = exports.euiOverlayMaskBodyStyles = void 0;

var _react = require("@emotion/react");

var _global_styling = require("../../global_styling");

var _services = require("../../services");

function _EMOTION_STRINGIFIED_CSS_ERROR__() { return "You have tried to stringify object returned from `css` function. It isn't supposed to be used directly (e.g. as value of the `className` prop), but rather handed to emotion so it can handle it (e.g. as value of `css` prop)."; }

var euiOverlayMaskStyles = function euiOverlayMaskStyles(_ref) {
  var euiTheme = _ref.euiTheme;
  return {
    euiOverlayMask: /*#__PURE__*/(0, _react.css)(".euiOverlayMask{position:fixed;", (0, _global_styling.logicalCSS)('top', 0), " ", (0, _global_styling.logicalCSS)('left', 0), " ", (0, _global_styling.logicalCSS)('right', 0), " ", (0, _global_styling.logicalCSS)('bottom', 0), " display:flex;align-items:center;justify-content:center;", (0, _global_styling.logicalCSS)('padding-bottom', '10vh'), ";animation:", _global_styling.euiAnimFadeIn, " ", euiTheme.animation.fast, " ease-in;background:", (0, _services.transparentize)(euiTheme.colors.ink, 0.5), ";};label:euiOverlayMask;"),
    aboveHeader: /*#__PURE__*/(0, _react.css)(".euiOverlayMask{z-index:", euiTheme.levels.mask, ";};label:aboveHeader;"),
    belowHeader: /*#__PURE__*/(0, _react.css)(".euiOverlayMask{z-index:", euiTheme.levels.maskBelowHeader, ";", (0, _global_styling.logicalCSS)('top', "".concat(euiTheme.base * 3, "px")), ";};label:belowHeader;")
  };
};

exports.euiOverlayMaskStyles = euiOverlayMaskStyles;
var euiOverlayMaskBodyStyles = process.env.NODE_ENV === "production" ? {
  name: "131160-euiOverlayMaskBodyStyles",
  styles: "body{overflow:hidden;};label:euiOverlayMaskBodyStyles;"
} : {
  name: "131160-euiOverlayMaskBodyStyles",
  styles: "body{overflow:hidden;};label:euiOverlayMaskBodyStyles;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};
exports.euiOverlayMaskBodyStyles = euiOverlayMaskBodyStyles;
