// @kbn/securitysolution-rules/target_web/src/rule_type_constants.js
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

/**
 * Id for the legacy siem signals alerting type
 */
export var SIGNALS_ID = "siem.signals";

/**
 * IDs for alerts-as-data rule types
 */
var RULE_TYPE_PREFIX = "siem";
export var EQL_RULE_TYPE_ID = "".concat(RULE_TYPE_PREFIX, ".eqlRule");
export var INDICATOR_RULE_TYPE_ID = "".concat(RULE_TYPE_PREFIX, ".indicatorRule");
export var ML_RULE_TYPE_ID = "".concat(RULE_TYPE_PREFIX, ".mlRule");
export var QUERY_RULE_TYPE_ID = "".concat(RULE_TYPE_PREFIX, ".queryRule");
export var SAVED_QUERY_RULE_TYPE_ID = "".concat(RULE_TYPE_PREFIX, ".savedQueryRule");
export var THRESHOLD_RULE_TYPE_ID = "".concat(RULE_TYPE_PREFIX, ".thresholdRule");
export var NEW_TERMS_RULE_TYPE_ID = "".concat(RULE_TYPE_PREFIX, ".newTermsRule");
