// @kbn/core-saved-objects-base-server-internal/target_node/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "LEGACY_URL_ALIAS_TYPE", {
  enumerable: true,
  get: function () {
    return _legacy_alias.LEGACY_URL_ALIAS_TYPE;
  }
});
Object.defineProperty(exports, "SavedObjectConfig", {
  enumerable: true,
  get: function () {
    return _saved_objects_config.SavedObjectConfig;
  }
});
Object.defineProperty(exports, "SavedObjectTypeRegistry", {
  enumerable: true,
  get: function () {
    return _saved_objects_type_registry.SavedObjectTypeRegistry;
  }
});
Object.defineProperty(exports, "SavedObjectsSerializer", {
  enumerable: true,
  get: function () {
    return _serialization.SavedObjectsSerializer;
  }
});
Object.defineProperty(exports, "SavedObjectsTypeValidator", {
  enumerable: true,
  get: function () {
    return _validation.SavedObjectsTypeValidator;
  }
});
Object.defineProperty(exports, "decodeRequestVersion", {
  enumerable: true,
  get: function () {
    return _version.decodeRequestVersion;
  }
});
Object.defineProperty(exports, "decodeVersion", {
  enumerable: true,
  get: function () {
    return _version.decodeVersion;
  }
});
Object.defineProperty(exports, "encodeHitVersion", {
  enumerable: true,
  get: function () {
    return _version.encodeHitVersion;
  }
});
Object.defineProperty(exports, "encodeVersion", {
  enumerable: true,
  get: function () {
    return _version.encodeVersion;
  }
});
Object.defineProperty(exports, "getIndexForType", {
  enumerable: true,
  get: function () {
    return _utils.getIndexForType;
  }
});
Object.defineProperty(exports, "getObjectKey", {
  enumerable: true,
  get: function () {
    return _utils.getObjectKey;
  }
});
Object.defineProperty(exports, "getProperty", {
  enumerable: true,
  get: function () {
    return _mappings.getProperty;
  }
});
Object.defineProperty(exports, "getRootProperties", {
  enumerable: true,
  get: function () {
    return _mappings.getRootProperties;
  }
});
Object.defineProperty(exports, "getRootPropertiesObjects", {
  enumerable: true,
  get: function () {
    return _mappings.getRootPropertiesObjects;
  }
});
Object.defineProperty(exports, "getTypes", {
  enumerable: true,
  get: function () {
    return _mappings.getTypes;
  }
});
Object.defineProperty(exports, "parseObjectKey", {
  enumerable: true,
  get: function () {
    return _utils.parseObjectKey;
  }
});
Object.defineProperty(exports, "savedObjectsConfig", {
  enumerable: true,
  get: function () {
    return _saved_objects_config.savedObjectsConfig;
  }
});
Object.defineProperty(exports, "savedObjectsMigrationConfig", {
  enumerable: true,
  get: function () {
    return _saved_objects_config.savedObjectsMigrationConfig;
  }
});
var _legacy_alias = require("./src/legacy_alias");
var _mappings = require("./src/mappings");
var _serialization = require("./src/serialization");
var _validation = require("./src/validation");
var _version = require("./src/version");
var _saved_objects_config = require("./src/saved_objects_config");
var _saved_objects_type_registry = require("./src/saved_objects_type_registry");
var _utils = require("./src/utils");
